Sources/NIOExtras/WritePCAPHandler.swift

```diff
@@ -808,6 +808,131 @@ extension NIOWritePCAPHandler {
+            case createNewPCAPFile
+        }
+        
+        public struct Error: Swift.Error {
+            public var errorCode: Int
+
+            internal enum ErrorCode: Int {
+                case cannotOpenFileError = 1
+                case cannotWriteToFileError
+            }
+        }
+        
+        private enum State {
+            case running
+            case error(Swift.Error)
+        }
+        
+        /// Creates an AsynchronizedFileSink for writing to a .pcap file at the given path.
+        /// If fileWritingMode is `.createNewPCAPFile`, a file header is written.
+        public static func fileSinkWritingToFile(
+            path: String,
+            fileWritingMode: FileWritingMode = .createNewPCAPFile,
+            errorHandler: @escaping (Swift.Error) -> Void,
+            on eventLoop: EventLoop
+        ) async throws -> AsynchronizedFileSink {
+            let oflag: CInt = fileWritingMode == .createNewPCAPFile ? (O_TRUNC | O_CREAT) : O_APPEND
+            let fd: CInt = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<CInt, Swift.Error>) in
+                path.withCString { pathPtr in
+                    let fd: Int32 = open(pathPtr, O_WRONLY | oflag, 0o600)
+                    if fd < 0 {
+                        continuation.resume(throwing: Error(errorCode: Error.ErrorCode.cannotOpenFileError.rawValue))
+                    } else {
+                        continuation.resume(returning: fd)
+                    }
+                }
+            }
+            
+            if fileWritingMode == .createNewPCAPFile {
+                _ = try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<CInt, Swift.Error>) in
+                    NIOWritePCAPHandler.pcapFileHeader.withUnsafeReadableBytes { ptr in
+                        let writeOk: Bool = sysWrite(fd, ptr.baseAddress, ptr.count) == ptr.count
+                        if writeOk {
+                            continuation.resume(returning: 0)  // Dummy value
+                        } else {
+                            continuation.resume(throwing: Error(errorCode: Error.ErrorCode.cannotWriteToFileError.rawValue))
+                        }
+                    }
+                }
+            }
+            let fileHandle: NIOFileHandle = NIOFileHandle(_deprecatedTakingOwnershipOfDescriptor: fd)
+            return AsynchronizedFileSink(fileHandle: fileHandle, eventLoop: eventLoop, errorHandler: errorHandler)
+        }
+        
+        private init(
+            fileHandle: NIOFileHandle,
+            eventLoop: EventLoop,
+            errorHandler: @escaping (Swift.Error) -> Void
+        ) {
+            self.fileHandle = fileHandle
+            self.eventLoop = eventLoop
+            self.errorHandler = errorHandler
+        }
+        
+        public func write(buffer: ByteBuffer) async throws {
+            try await self.eventLoop.submit {
```

Comment:
This still blocks the event loop. By being `async`, this method will already run off the event loop, so we don't need to do anything other than issue the write call.